File: testcases/c/fantasy_test.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <assert.h>
#include "namegen.h"  /* Must declare:
                        - int namegen(char *dst, unsigned long len, const char *pattern, unsigned long *seed);
                        - Return codes: NAMEGEN_SUCCESS, NAMEGEN_TRUNCATED, NAMEGEN_INVALID, NAMEGEN_TOO_DEEP
                        - NAMEGEN_MAX_DEPTH */

#define BUF_SIZE 256

static void test_literal(void) {
    char output[BUF_SIZE];
    unsigned long seed = 12345;
    int ret = namegen(output, BUF_SIZE, "Hello", &seed);
    assert(ret == NAMEGEN_SUCCESS);
    assert(strcmp(output, "Hello") == 0);
    printf("test_literal passed: \"%s\"\n", output);
}

static void test_group_literal(void) {
    char output[BUF_SIZE];
    unsigned long seed = 54321;
    int ret = namegen(output, BUF_SIZE, "(Hello)", &seed);
    assert(ret == NAMEGEN_SUCCESS);
    assert(strcmp(output, "Hello") == 0);
    printf("test_group_literal passed: \"%s\"\n", output);
}

static void test_capitalization(void) {
    char output[BUF_SIZE];
    unsigned long seed = 11111;
    int ret = namegen(output, BUF_SIZE, "!(hello)", &seed);
    assert(ret == NAMEGEN_SUCCESS);
    assert(strcmp(output, "Hello") == 0);
    printf("test_capitalization passed: \"%s\"\n", output);
}

static void test_substitution_repeatability(void) {
    char output1[BUF_SIZE], output2[BUF_SIZE];
    unsigned long seed1 = 1, seed2 = 1;
    int ret1 = namegen(output1, BUF_SIZE, "s", &seed1);
    int ret2 = namegen(output2, BUF_SIZE, "s", &seed2);
    assert(ret1 == NAMEGEN_SUCCESS);
    assert(ret2 == NAMEGEN_SUCCESS);
    assert(strcmp(output1, output2) == 0);
    printf("test_substitution_repeatability passed: \"%s\"\n", output1);
}

static void test_choice(void) {
    char output[BUF_SIZE];
    unsigned long seed = 98765;
    int ret = namegen(output, BUF_SIZE, "(foo|bar)", &seed);
    assert(ret == NAMEGEN_SUCCESS);
    int isFoo = (strcmp(output, "foo") == 0);
    int isBar = (strcmp(output, "bar") == 0);
    assert(isFoo || isBar);
    printf("test_choice passed: \"%s\"\n", output);
}

static void test_invalid(void) {
    char output[BUF_SIZE];
    unsigned long seed = 123;
    int ret = namegen(output, BUF_SIZE, "(", &seed);
    assert(ret == NAMEGEN_INVALID);
    printf("test_invalid passed\n");
}

static void test_too_deep(void) {
    char pattern[512];
    int i;
    for (i = 0; i < NAMEGEN_MAX_DEPTH + 1; i++) {
        pattern[i] = '(';
    }
    pattern[i] = '\0';
    char output[BUF_SIZE];
    unsigned long seed = 555;
    int ret = namegen(output, BUF_SIZE, pattern, &seed);
    assert(ret == NAMEGEN_TOO_DEEP);
    printf("test_too_deep passed\n");
}

static void test_truncation(void) {
    char output[6]; // Very small buffer.
    unsigned long seed = 777;
    int ret = namegen(output, sizeof(output), "HelloWorld", &seed);
    assert(ret == NAMEGEN_TRUNCATED);
    printf("test_truncation passed: output = \"%s\"\n", output);
}



int main(void) {
    test_literal();
    test_group_literal();
    test_capitalization();
    test_substitution_repeatability();
    test_choice();
    test_invalid();
    test_too_deep();
    test_truncation();
    printf("All namegen tests passed.\n");
    return 0;
}

File: testcases/rust/fantasy_test.rs
// Assume these definitions in your Rust port:
#[derive(Debug, PartialEq)]
pub enum NamegenResult {
    Success,
    Truncated,
    Invalid,
    TooDeep,
}

pub const NAMEGEN_MAX_DEPTH: usize = 32;

// The namegen function signature:
pub fn namegen(dst: &mut [u8], pattern: &str, seed: &mut u64) -> NamegenResult {
    // Your implementation goes here...
    unimplemented!()
}

// Helper: Convert a null-terminated buffer to a &str.
fn buffer_to_str(buf: &[u8]) -> &str {
    let len = buf.iter().position(|&c| c == 0).unwrap_or(buf.len());
    std::str::from_utf8(&buf[..len]).unwrap()
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_literal() {
        let mut buf = [0u8; 256];
        let mut seed = 12345u64;
        let res = namegen(&mut buf, "Hello", &mut seed);
        assert_eq!(res, NamegenResult::Success);
        assert_eq!(buffer_to_str(&buf), "Hello");
        println!("test_literal passed: output = \"{}\"", buffer_to_str(&buf));
    }

    #[test]
    fn test_group_literal() {
        let mut buf = [0u8; 256];
        let mut seed = 54321u64;
        let res = namegen(&mut buf, "(Hello)", &mut seed);
        assert_eq!(res, NamegenResult::Success);
        assert_eq!(buffer_to_str(&buf), "Hello");
        println!("test_group_literal passed: output = \"{}\"", buffer_to_str(&buf));
    }

    #[test]
    fn test_capitalization() {
        let mut buf = [0u8; 256];
        let mut seed = 11111u64;
        let res = namegen(&mut buf, "!(hello)", &mut seed);
        assert_eq!(res, NamegenResult::Success);
        assert_eq!(buffer_to_str(&buf), "Hello");
        println!("test_capitalization passed: output = \"{}\"", buffer_to_str(&buf));
    }

    #[test]
    fn test_substitution_repeatability() {
        let mut buf1 = [0u8; 256];
        let mut buf2 = [0u8; 256];
        let mut seed1 = 1u64;
        let mut seed2 = 1u64;
        let res1 = namegen(&mut buf1, "s", &mut seed1);
        let res2 = namegen(&mut buf2, "s", &mut seed2);
        assert_eq!(res1, NamegenResult::Success);
        assert_eq!(res2, NamegenResult::Success);
        let out1 = buffer_to_str(&buf1);
        let out2 = buffer_to_str(&buf2);
        assert_eq!(out1, out2);
        println!("test_substitution_repeatability passed: output = \"{}\"", out1);
    }

    #[test]
    fn test_choice() {
        let mut buf = [0u8; 256];
        let mut seed = 98765u64;
        let res = namegen(&mut buf, "(foo|bar)", &mut seed);
        assert_eq!(res, NamegenResult::Success);
        let out = buffer_to_str(&buf);
        assert!(out == "foo" || out == "bar");
        println!("test_choice passed: output = \"{}\"", out);
    }

    #[test]
    fn test_invalid() {
        let mut buf = [0u8; 256];
        let mut seed = 123u64;
        let res = namegen(&mut buf, "(", &mut seed);
        assert_eq!(res, NamegenResult::Invalid);
        println!("test_invalid passed");
    }

    #[test]
    fn test_too_deep() {
        let pattern: String = std::iter::repeat("(")
            .take(NAMEGEN_MAX_DEPTH + 1)
            .collect();
        let mut buf = [0u8; 256];
        let mut seed = 555u64;
        let res = namegen(&mut buf, &pattern, &mut seed);
        assert_eq!(res, NamegenResult::TooDeep);
        println!("test_too_deep passed");
    }

    #[test]
    fn test_truncation() {
        let mut buf = [0u8; 6];
        let mut seed = 777u64;
        let res = namegen(&mut buf, "HelloWorld", &mut seed);
        assert_eq!(res, NamegenResult::Truncated);
        println!("test_truncation passed: output = \"{}\"", buffer_to_str(&buf));
    }

    #[test]
    fn test_complex_pattern() {
        let mut buf = [0u8; 256];
        let mut seed = 24680u64;
        // Use updated pattern "!(c)ast (m)onster"
        let res = namegen(&mut buf, "!(c)ast (m)onster", &mut seed);
        assert_eq!(res, NamegenResult::Success);
        let out = buffer_to_str(&buf);
        // Expected output is "Cast monster"
        assert_eq!(out, "Cast monster");
        println!("test_complex_pattern passed: output = \"{}\"", out);
    }
}
